### PR TITLE
Issue 1419 - configurable memory and CPU limits for a service

### PR DESCRIPTION
**Pull request by sirotsinskuy**
Signed-off-by: Antonii Sirotsinskyi <asiro@softserveinc.com>

### CODE REVIEW

File: cli/native_deployment/native_deployment.go [hunk 1]
```diff
@@ -159,7 +159,7 @@ func (p *NativeDeploymentConfigPlugin) Validate(dep interface{}, cdep interface{
 }
 
 // This can't be a const because a map literal isn't a const in go
-var VALID_DEPLOYMENT_FIELDS = map[string]int8{"image": 1, "privileged": 1, "cap_add": 1, "environment": 1, "devices": 1, "binds": 1, "specific_ports": 1, "command": 1, "ports": 1, "ephemeral_ports": 1, "tmpfs": 1, "network": 1}
+var VALID_DEPLOYMENT_FIELDS = map[string]int8{"image": 1, "privileged": 1, "cap_add": 1, "environment": 1, "devices": 1, "binds": 1, "specific_ports": 1, "command": 1, "ports": 1, "ephemeral_ports": 1, "tmpfs": 1, "network": 1, "max_memory_kb": 1, "max_cpus": 1}
```
Comment (dabooz) [inline cli/native_deployment/native_deployment.go:162]: Why did you choose KB for memory size. I think allowing the user to specify in MB is more usable.

File: docs/deployment_string.md [hunk 2]
```diff
@@ -35,6 +35,8 @@ Because Horizon uses the Docker API to start the containers on an edge node, man
     - `ephemeral_ports`: `[{"localhost_only":true, "port_and_protocol":"7777/udp"}, {"port_and_protocol":"8888"}...]` - publish a container port to an ephemeral host port. If `localhost_only` is set to true, the localhost ip address (`127.0.0.1`) will be used as the host network interface this port should listen on. Otherwise, all the host network interfaces on the host will be listened by this port. If the protocol is not specified after the port number for `port_and_protocol`, it defaults to `tcp`.
     - `command`: `["--myfirstarg","argvalue",...]` - override the start CMD specified the dockerfile, or append to the ENTRYPOINT specified in the dockerfile.
     - `network`: `"host"` - start the container with host network mode. When network is set to host, the service can only be deployed to nodes with property openhorizon.allowPrivileged set to true.
+    - `max_memory_kb`: `"4096"` - the maximum amount of memory the service's container can use
```
Comment (dabooz) [inline docs/deployment_string.md:38]: "the maximum amount of memory in KB that the service's container can use." Obviously change this to MB if we decide to change the field units to MB.

File: containermessage/models.go [hunk 3]
```diff
@@ -151,6 +151,8 @@ type Service struct {
 	Ports            []docker.PortBinding `json:"ports,omitempty"`
 	EphemeralPorts   []Port               `json:"ephemeral_ports,omitempty"`
 	SpecificPorts    []docker.PortBinding `json:"specific_ports,omitempty"` // obselete. for backward compatibility only, new way should use ports instead.
+	MaxMemoryKb      int64                `json:"max_memory_kb,omitempty"`
```
Comment (dabooz) [inline containermessage/models.go:154]: Please update one of the services created by e2edev to specify a memory and cpu limit. And then add code to verify that docker inspect of that service (when it's running) shows the cpu and memory limits you specified.

**Review by dabooz — CHANGES_REQUESTED** [review]

File: containermessage/models.go (hunk 3 above)
Comment (sirotsinskuy) [reply containermessage/models.go]: Updated PR with additional memory&cpu check

File: docs/deployment_string.md [hunk 4]
```diff
@@ -35,6 +35,8 @@ Because Horizon uses the Docker API to start the containers on an edge node, man
     - `ephemeral_ports`: `[{"localhost_only":true, "port_and_protocol":"7777/udp"}, {"port_and_protocol":"8888"}...]` - publish a container port to an ephemeral host port. If `localhost_only` is set to true, the localhost ip address (`127.0.0.1`) will be used as the host network interface this port should listen on. Otherwise, all the host network interfaces on the host will be listened by this port. If the protocol is not specified after the port number for `port_and_protocol`, it defaults to `tcp`.
     - `command`: `["--myfirstarg","argvalue",...]` - override the start CMD specified the dockerfile, or append to the ENTRYPOINT specified in the dockerfile.
     - `network`: `"host"` - start the container with host network mode. When network is set to host, the service can only be deployed to nodes with property openhorizon.allowPrivileged set to true.
+    - `max_memory_mb`: `"4096"` - the maximum amount of memory the service's container can use
```
Comment (dabooz) [inline docs/deployment_string.md:38]: These values should not be quoted.

**Review by dabooz — CHANGES_REQUESTED** [review]

**Review by dabooz — APPROVED** [review]

**Pull request merged**